Periodically-Syncing-HyperChains.md

```diff
@@ -540,6 +553,255 @@ During the block production epoch, blocks are considered valid only if they are
+      - The arguments are:
+        - **Epoch**: The epoch number for which the finalization is being done.
+        - **Chosen Fork**: The block hash of the chosen fork.
+        - **new_epoch_length**: The length of the next epoch calculated by multiplying submitted deltas
+                                by staking percentage rounded towards 0 + current length.
+        - **pc_root_hash**: The root hash at PC the given PC height that is used as seed for leader     election.
+        - **Validator**: The address of the validator creating the finalization transaction.
+        - **Votes Proof**: A list of votes from other validators, each containing their transaction payloads.
+          - The block hash they voted for.
+          - The epoch length deltas.
+          - The validator’s address.
+          - The validator’s signature.
+      - The call is obviously signed by validator creating the finalization transaction to ensure authenticity, as with any transaction/contract call. This is the same validator that is
+      producing the block so the transaction will not be refused. A correct call should give a reward.
+      an illegal call can be challenged and result in a penalty.
+
+      By including the votes of other validators the call serves as verifiable proof that a quorum has been reached.
+      - This call is recorded in the final block of the epoch.
+
+   3. **Updating Local States**:
+      - Upon verifying the "Finalize" transaction, all validators update their local state to reflect the chosen fork as the correct chain.
+      - The network then proceeds to the next epoch based on this agreed-upon state.
+
+#### Timeouts
+
+To implement a robust BFT voting mechanism, it's essential to establish clear rules for timeouts, handle situations where a majority is not reached, and address scenarios where a validator might ignore some votes and create a minority vote in the finalization. Here's how we can approach these challenges:
+
+##### Setting Timeouts
+
+We can define timeouts for each phase of the voting process. These timeouts should be possible to configure
+(within some bounds) when initializing a hyperchain.
+
+- **Proposal Timeout**: A predefined period (e.g., 10 seconds) within which validators can submit their fork proposals. After this period, no new proposals are accepted.
+
+- **Voting Timeout**: A defined period (e.g., 30 seconds) for validators to submit their votes for a fork proposal. This time window allows all validators to observe the proposals and cast their votes. The timeout duration can be set based on the expected network latency and block production time.
+
+- **Finalization Timeout**: A specified period (e.g., 20 seconds) after the voting phase ends, within which a validator must create and broadcast the "Finalize" transaction. If no finalization occurs within this period, the network takes predefined corrective actions. (The new leader in the next epoch just runs with his preferred fork.)
```

Comment:
I guess these are max/timeouts - i.e. in the happy case it goes much faster? But doesn't this uncertainty make epoch lenght adjustments a bit hit or miss... If we have a normal slot-length of 3 seconds then the last slot can be up to 20 times longer with these (rather reasonable) numbers 🤔

Reply:
Yes, we need to figure something out here. We also need some solid measurements to get good suggested timeouts.
One solution could be to automatically adjust the block length of the next epoch during a slow finalizing process.

Reply:
> Yes, we need to figure something out here. We also need some solid measurements to get good suggested timeouts. One solution could be to automatically adjust the block length of the next epoch during a slow finalizing process.

That sounds dangerous... At that point you know the random seed, so if you want to produce the last block in the epoch you just need to delay the finalization long enough 🤔

Reply:
Yes. It's better to take the max or 2/3 of the max time and allow the last block always to be slow. Or rather, we set a schedule for when (in NTP-agreed wall clock time) the last block of an epoch (or possibly first block of next epoch) should be produced and try to follow that.

Reply:
It could be possible, though potentially tricky/confusing to let the last block of the epoch contain _multiple_ micro-blocks to avoid the hick-up in transaction inclusion 🤔

Reply:
> last block of the epoch contain _multiple_ micro-blocks

would there otherwise only be one? Otherwise that wouldn't be to bad and if we expect for `if(blockheight == (blockheight % epochsize == 0))` to have additional 30 sec for the finalization process

Reply:
Yes, the idea (not sure if it is written down) was to have just one micro-block and a key-block in each validator slot. Another trick there is to have the micro-block _before_ the key-block in order to efficiently signal if there is no micro-block in a slot.